contracts/cw721-base/src/execute.rs

```diff
@@ -125,6 +123,39 @@ where
             .add_attribute("owner", owner)
             .add_attribute("token_id", token_id))
     }
+
+    pub fn update_ownership(
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        action: cw_ownable::Action,
+    ) -> Result<Response<C>, ContractError> {
+        let ownership = cw_ownable::update_ownership(deps, &env.block, &info.sender, action)?;
+        Ok(Response::new().add_attributes(ownership.into_attributes()))
+    }
+
+    pub fn migrate(deps: DepsMut, msg: MigrateMsg) -> Result<Response<C>, ContractError> {
+        let ContractVersion { version, .. } = get_contract_version(deps.storage)?;
+        set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+        match msg {
```

Comment:
Against this approach, because this way the contract will contain all the historical migration logics: e.g. a v0.18 contract will include both v0.17 -> v0.18 migration logics, but also that for v0.16 -> v0.17, etc. This way, the wasm binary size will keep growing as we add more and more migration logics to it.

Imo, migration logics should be included in a way that older, unused logics can be optimized away. Can you take a look at what I do in badges project and let me know what you think: https://github.com/steak-enjoyers/badges/tree/master/contracts/hub/src/upgrades

Reply:
thanks for the thoughts, i read your earlier comment but already had this done, minus test updates. here's my argument for this approach:

what is our biggest concern for contract migrations? imo:

1. an accidental state key change in a newer version that causes state to go missing after migration. [ex](https://github.com/DA0-DA0/dao-contracts/pull/548#discussion_r1029936817).
2. a small change in the way types are serialized between versions (perhaps in a dependency) causes deserialization to fail post-migration. [ex](https://github.com/DA0-DA0/dao-contracts/pull/465#discussion_r958808276).

as the examples show, careful people mess these things up frequently.

in my view, this system is optimal then as it protects us against these in many ways.

1. we are guaranteed to load from the same hardcoded state key.
2. we are guaranteed to get the deserialization method used by the old version.

then, using the old import we can run unit tests against _exactly_ the old contract and perform state checks before and after (see the `mint_transfer_burn` call before and after migration in my integration test). we track contract size and upload cost over time in dao-contracts and i have not seen notable increases from adding these dependencies, i expect this is the rust linker removing the unused parts.

i do understand that this reduces the ease of multi-hop migrations. in dao-contracts we only maintain a `FromVersion {}` migration variant for one version. So, do do a multi-hop upgrade you'd need to migrate twice.

Reply:
Understand the concern, but my opinion is this: _each version of the contract should only contain one migration logic, from the immediate previous version to the current version._ E.g. contract 0.17 should only contain 0.16->0.17, contract 0.18 should only contain 0.17->0.18, so on.

Of course, when migrating 0.16->0.17, the contract must check cw2 that it's indeed being migrated from 0.16.

If you have a 0.16 contract and want to migrate to 0.18, you have to do this sequentially: first 0.16->0.17 using the 0.18 code_id, then 0.17->0.18 with the 0.18 code_id.

With this design we get both a small wasm binary size, and avoid the issue you're concerned about.

Reply:
@larry0x would you mind taking a look at this commit and seeing if it fits what you're saying? https://github.com/CosmWasm/cw-nfts/pull/109/commits/6b934b624353826afb706b0604c79b485936d53a